File: Controle de Pagamentos/C#/Unit1.cs
using MySql.Data.MySqlClient;
using System;
using System.Collections.Generic;
using System.Windows.Forms;
using Util;
using VO;

namespace Seguranca
{
	
	  
	public partial class Form1 : Form
	{

        private static List<ControlePagamentoVO> ListaPagamento;

        private MySqlConnection conexao;
        private MySqlCommand comando;
        private MySqlDataReader leitor;
        public string ConsultaSQL;
		
		public Form1()
		{
			// Required for Windows Form Designer support
			InitializeComponent();
            Grid1.AutoGenerateColumns = false;
            Grid2.AutoGenerateColumns = false;
        }


        #region Controle via Banco
        private void Button2_Click(object sender, System.EventArgs e)
        {
            ConsultaSQL = "select * from controle_pagamento";
            try
            {
                conexao = dbConnection.conectar();

                ListaPagamento = new List<ControlePagamentoVO>();

                comando = new MySqlCommand(ConsultaSQL, conexao);
                leitor = comando.ExecuteReader();
                while (leitor.Read())
                {
                    ControlePagamentoVO Pagamento = new ControlePagamentoVO();

                    Pagamento.Id = Convert.ToInt32(leitor["id"]);
                    Pagamento.a = Convert.ToString(leitor["a"]);
                    Pagamento.b = Convert.ToString(leitor["b"]);
                    Pagamento.c = Convert.ToString(leitor["c"]);
                    Pagamento.d = Convert.ToString(leitor["d"]);
                    Pagamento.e = Convert.ToString(leitor["e"]);
                    Pagamento.f = Convert.ToString(leitor["f"]);

                    ListaPagamento.Add(Pagamento);
                }

                Grid1.DataSource = ListaPagamento;
            }
            catch (Exception eError)
            {
                MessageBox.Show(eError.Message);
                ListaPagamento = null;
            }
            finally
            {
                if (leitor != null)
                    leitor.Close();
            }
        }

        private void Button1_Click(object sender, EventArgs e)
        {
            ConsultaSQL = "insert into controle_pagamento ( " +
                                                    "a,b,c,d,e,f)" +
                                                    " values (" +
                                                    "?pa,?pb,?pc,?pd,?pe,?pf)";
            try
            {
                conexao = dbConnection.conectar();
                comando = new MySqlCommand(ConsultaSQL, conexao);

                comando.Parameters.AddWithValue("?pa", Biblioteca.Encripta(EditMes.Text));
                comando.Parameters.AddWithValue("?pb", Biblioteca.Encripta(EditAno.Text));
                comando.Parameters.AddWithValue("?pc", Biblioteca.Encripta(EditVencimento.Text));
                comando.Parameters.AddWithValue("?pd", Biblioteca.Encripta(EditBloqueio.Text));
                comando.Parameters.AddWithValue("?pe", Biblioteca.Encripta(EditLiberado.Text));

                comando.Parameters.AddWithValue("?pf", Biblioteca.MD5String(EditMes.Text + EditAno.Text + EditVencimento.Text + EditBloqueio.Text + EditLiberado.Text));
                
                comando.ExecuteNonQuery();
            }
            catch (Exception eError)
            {
                MessageBox.Show(eError.Message);
            }
            finally
            {
                if (leitor != null)
                    leitor.Close();
            }
        }

        private void Button3_Click(object sender, EventArgs e)
        {
            try
            {
                Memo1.Clear();
                Memo1.AppendText("Mês = " + Biblioteca.Desencripta(ListaPagamento[Grid1.CurrentRow.Index].a));
                Memo1.AppendText("\rAno = " + Biblioteca.Desencripta(ListaPagamento[Grid1.CurrentRow.Index].b));
                Memo1.AppendText("\rVencimento = " + Biblioteca.Desencripta(ListaPagamento[Grid1.CurrentRow.Index].c));
                Memo1.AppendText("\rBloqueio = " + Biblioteca.Desencripta(ListaPagamento[Grid1.CurrentRow.Index].d));
                Memo1.AppendText("\rLiberado = " + Biblioteca.Desencripta(ListaPagamento[Grid1.CurrentRow.Index].e));
            }
            catch (Exception eError)
            {
                MessageBox.Show(eError.Message);
            }
        }


        private void Button4_Click(object sender, EventArgs e)
        {
            try
            {
                string HashRegistro = Biblioteca.MD5String(
                                                            Biblioteca.Desencripta(ListaPagamento[Grid1.CurrentRow.Index].a) +
                                                            Biblioteca.Desencripta(ListaPagamento[Grid1.CurrentRow.Index].b) +
                                                            Biblioteca.Desencripta(ListaPagamento[Grid1.CurrentRow.Index].c) +
                                                            Biblioteca.Desencripta(ListaPagamento[Grid1.CurrentRow.Index].d) +
                                                            Biblioteca.Desencripta(ListaPagamento[Grid1.CurrentRow.Index].e)
                                                            );

                Memo1.AppendText("\r===================================");
                Memo1.AppendText("\rId = " + ListaPagamento[Grid1.CurrentRow.Index].Id);

                if (HashRegistro == ListaPagamento[Grid1.CurrentRow.Index].f)
                {
                    Memo1.AppendText("\rHash OK - Registro não violado.");
                }
                else
                {
                    Memo1.AppendText("\rHash Diferente - Registro violado.");
                }
                Memo1.AppendText("\r===================================");
            }
            catch (Exception eError)
            {
                MessageBox.Show(eError.Message);
            }
        }

        private void Button5_Click(object sender, EventArgs e)
        {
            bool Resultado = false;
            
            //  verifica se houve alguma tentativa de burlar os dados no banco de dados
            for (int i = 0; i <= ListaPagamento.Count - 1; i++)
            {
                Grid1.Focus();
                Grid1.CurrentCell = Grid1.Rows[i].Cells[0]; 
                Button4.PerformClick();
            }


            //  verifica se existe algum registro para o mês e para o ano - caso não exista, bloqueia o sistema.
            //  select * from tabela where mes=mes_atual and ano=ano_atual
            for (int i = 0; i <= ListaPagamento.Count - 1; i++)
            {
                if ((DateTime.Now.Year == int.Parse(Biblioteca.Desencripta(ListaPagamento[Grid1.CurrentRow.Index].b))) && (DateTime.Now.Month == int.Parse(Biblioteca.Desencripta(ListaPagamento[Grid1.CurrentRow.Index].a))))
                {
                    Resultado = true;
                    break;
                }
            }
            if (!Resultado)
            {
                MessageBox.Show("Não encontrou um registro com o mês e o ano atual... Sistema deve ser bloqueado.");
                return;
            }


            //  caso tenha encontrado o mes e ano, verifica se a data atual É maior que a data do vencimento
            if (DateTime.Now.Day > int.Parse(Biblioteca.Desencripta(ListaPagamento[Grid1.CurrentRow.Index].c)))
            {
                MessageBox.Show("A data atual é maior que a data do vencimento. Será que o cliente já pagou?");

                //  verifica se o cliente já realizou o pagamento através do campo Liberado
                if (Biblioteca.Desencripta(ListaPagamento[Grid1.CurrentRow.Index].e) == "N")
                {
                    MessageBox.Show("O cliente não pagou. Será que o sistema deve ser bloqueado?");

                    //  caso a data atual seja maior ou igual ? data definida para bloqueio, o sistema deve ser bloqueado
                    if (DateTime.Now.Day > int.Parse(Biblioteca.Desencripta(ListaPagamento[Grid1.CurrentRow.Index].d)))
                    {
                        MessageBox.Show("O sistema deve ser bloqueado. Informar o seguinte código para SH: \r\r" + ListaPagamento[Grid1.CurrentRow.Index].f);
                    }
                }
                else
                    MessageBox.Show("O cliente já pagou!");
            }

            //  Exercício: e se o cliente deixar de pagar durante dois meses?
        }

        private void Button6_Click(object sender, EventArgs e)
        {
            Button3.PerformClick();
            // 
            Memo1.AppendText("\r=====================================================");
            Memo1.AppendText("\rHash Registro Bloqueado  = " + ListaPagamento[Grid1.CurrentRow.Index].f);
            Memo1.AppendText("\rChave Desbloqueio = " + Biblioteca.MD5String(ListaPagamento[Grid1.CurrentRow.Index].f + "uma_chave"));
            Memo1.AppendText("\r=====================================================");
        }

        private void Button7_Click(object sender, EventArgs e)
        {
            if (EditChaveDesbloqueio.Text == Biblioteca.MD5String(ListaPagamento[Grid1.CurrentRow.Index].f + "uma_chave"))
            {
                // Exercício: qual o equívoco dessa instrução?
                ConsultaSQL = "update controle_pagamento set e = " + Biblioteca.QuotedStr( Biblioteca.Encripta("S") );
                try
                {
                    conexao = dbConnection.conectar();
                    comando = new MySqlCommand(ConsultaSQL, conexao);
                    comando.ExecuteNonQuery();
                }
                catch (Exception eError)
                {
                    MessageBox.Show(eError.Message);
                }
                finally
                {
                    if (leitor != null)
                        leitor.Close();
                }

                MessageBox.Show("Dados Atualizados. Sistema Liberado!");
            }
            else
                MessageBox.Show("Chave não confere - entre em contato com a SH.");

        }
        #endregion

		#region Controle via Arquivo
        private void Button10_Click(object sender, EventArgs e)
        {
            // Aqui os dados estão sendo carregados a partir do arquivo
            ListaPagamento = ControlePagamentoVO.ListaDoArquivo("fake.dll");
            Grid2.DataSource = ListaPagamento;

            // Exercício: implemente a inclusão dos dados no arquivo
        }

        private void Button11_Click(object sender, EventArgs e)
        {
            try
            {
                Memo2.Clear();
                Memo2.AppendText("Mês = " + Biblioteca.Desencripta(ListaPagamento[Grid2.CurrentRow.Index].a));
                Memo2.AppendText("\rAno = " + Biblioteca.Desencripta(ListaPagamento[Grid2.CurrentRow.Index].b));
                Memo2.AppendText("\rVencimento = " + Biblioteca.Desencripta(ListaPagamento[Grid2.CurrentRow.Index].c));
                Memo2.AppendText("\rBloqueio = " + Biblioteca.Desencripta(ListaPagamento[Grid2.CurrentRow.Index].d));
                Memo2.AppendText("\rLiberado = " + Biblioteca.Desencripta(ListaPagamento[Grid2.CurrentRow.Index].e));
            }
            catch (Exception eError)
            {
                MessageBox.Show(eError.Message);
            }
        }

        private void Button12_Click(object sender, EventArgs e)
        {
            try
            {
                string HashRegistro = Biblioteca.MD5String(
                                                            Biblioteca.Desencripta(ListaPagamento[Grid2.CurrentRow.Index].a) +
                                                            Biblioteca.Desencripta(ListaPagamento[Grid2.CurrentRow.Index].b) +
                                                            Biblioteca.Desencripta(ListaPagamento[Grid2.CurrentRow.Index].c) +
                                                            Biblioteca.Desencripta(ListaPagamento[Grid2.CurrentRow.Index].d) +
                                                            Biblioteca.Desencripta(ListaPagamento[Grid2.CurrentRow.Index].e)
                                                            );

                Memo2.AppendText("\r===================================");
                Memo2.AppendText("\rId = " + ListaPagamento[Grid2.CurrentRow.Index].Id);

                if (HashRegistro == ListaPagamento[Grid2.CurrentRow.Index].f)
                {
                    Memo2.AppendText("\rHash OK - Registro não violado.");
                }
                else
                {
                    Memo2.AppendText("\rHash Diferente - Registro violado.");
                }
                Memo2.AppendText("\r===================================");
            }
            catch (Exception eError)
            {
                MessageBox.Show(eError.Message);
            }
        }

        private void Button13_Click(object sender, EventArgs e)
        {
            bool Resultado = false;

            //  verifica se houve alguma tentativa de burlar os dados no banco de dados
            for (int i = 0; i <= ListaPagamento.Count - 1; i++)
            {
                Grid2.Focus();
                Grid2.CurrentCell = Grid2.Rows[i].Cells[0];
                Button12.PerformClick();
            }


            //  verifica se existe algum registro para o mês e para o ano - caso não exista, bloqueia o sistema.
            //  select * from tabela where mes=mes_atual and ano=ano_atual
            for (int i = 0; i <= ListaPagamento.Count - 1; i++)
            {
                if ((DateTime.Now.Year == int.Parse(Biblioteca.Desencripta(ListaPagamento[Grid2.CurrentRow.Index].b))) && (DateTime.Now.Month == int.Parse(Biblioteca.Desencripta(ListaPagamento[Grid2.CurrentRow.Index].a))))
                {
                    Resultado = true;
                    break;
                }
            }
            if (!Resultado)
            {
                MessageBox.Show("Não encontrou um registro com o mês e o ano atual... Sistema deve ser bloqueado.");
                return;
            }


            //  caso tenha encontrado o mes e ano, verifica se a data atual É maior que a data do vencimento
            if (DateTime.Now.Day > int.Parse(Biblioteca.Desencripta(ListaPagamento[Grid2.CurrentRow.Index].c)))
            {
                MessageBox.Show("A data atual é maior que a data do vencimento. Será que o cliente já pagou?");

                //  verifica se o cliente já realizou o pagamento através do campo Liberado
                if (Biblioteca.Desencripta(ListaPagamento[Grid2.CurrentRow.Index].e) == "N")
                {
                    MessageBox.Show("O cliente não pagou. Será que o sistema deve ser bloqueado?");

                    //  caso a data atual seja maior ou igual ? data definida para bloqueio, o sistema deve ser bloqueado
                    if (DateTime.Now.Day > int.Parse(Biblioteca.Desencripta(ListaPagamento[Grid2.CurrentRow.Index].d)))
                    {
                        MessageBox.Show("O sistema deve ser bloqueado. Informar o seguinte código para SH: \r\r" + ListaPagamento[Grid2.CurrentRow.Index].f);
                    }
                }
                else
                    MessageBox.Show("O cliente já pagou!");
            }

            //  Exercício: e se o cliente deixar de pagar durante dois meses?
        }

        private void Button14_Click(object sender, EventArgs e)
        {
            Button11.PerformClick();
            // 
            Memo2.AppendText("\r=====================================================");
            Memo2.AppendText("\rHash Registro Bloqueado  = " + ListaPagamento[Grid2.CurrentRow.Index].f);
            Memo2.AppendText("\rChave Desbloqueio = " + Biblioteca.MD5String(ListaPagamento[Grid2.CurrentRow.Index].f + "uma_chave"));
            Memo2.AppendText("\r=====================================================");
        }

        private void Button8_Click(object sender, EventArgs e)
        {
            // Exercício: altere o arquivo no campo Liberado="S" criptografado
        }
        #endregion


    }
	
	
}

File: Controle de Pagamentos/Delphi/Unit1.pas
unit Unit1;

interface

uses
  Winapi.Windows, Winapi.Messages, System.SysUtils, System.Variants, System.Classes, Vcl.Graphics,
  Vcl.Controls, Vcl.Forms, Vcl.Dialogs, Vcl.ExtCtrls, Vcl.StdCtrls,
  Vcl.ComCtrls, Data.DBXMySQL, Data.DB, Data.SqlExpr, Vcl.Grids, Vcl.DBGrids, Data.FMTBcd,
  Datasnap.Provider, Datasnap.DBClient, DateUtils;

type
  TForm1 = class(TForm)
    Panel1: TPanel;
    PageControl1: TPageControl;
    TabSheet1: TTabSheet;
    TabSheet2: TTabSheet;
    Conexao: TSQLConnection;
    Button1: TButton;
    DBGrid1: TDBGrid;
    EditMes: TLabeledEdit;
    EditAno: TLabeledEdit;
    EditVencimento: TLabeledEdit;
    EditBloqueio: TLabeledEdit;
    EditLiberado: TLabeledEdit;
    Button2: TButton;
    SQLTable1: TSQLTable;
    DataSource1: TDataSource;
    ClientDataSet1: TClientDataSet;
    DataSetProvider1: TDataSetProvider;
    SQLTable1id: TIntegerField;
    SQLTable1a: TStringField;
    SQLTable1b: TStringField;
    SQLTable1c: TStringField;
    SQLTable1d: TStringField;
    SQLTable1e: TStringField;
    ClientDataSet1id: TIntegerField;
    ClientDataSet1a: TStringField;
    ClientDataSet1b: TStringField;
    ClientDataSet1c: TStringField;
    ClientDataSet1d: TStringField;
    ClientDataSet1e: TStringField;
    SQLTable1f: TStringField;
    ClientDataSet1f: TStringField;
    Memo1: TMemo;
    Button3: TButton;
    Button4: TButton;
    Button5: TButton;
    Button6: TButton;
    Button7: TButton;
    EditChaveDesbloqueio: TLabeledEdit;
    ClientDataSet2: TClientDataSet;
    ClientDataSet2a: TStringField;
    DBGrid2: TDBGrid;
    DataSource2: TDataSource;
    ClientDataSet2id: TIntegerField;
    ClientDataSet2b: TStringField;
    ClientDataSet2c: TStringField;
    ClientDataSet2d: TStringField;
    ClientDataSet2e: TStringField;
    ClientDataSet2f: TStringField;
    Memo2: TMemo;
    EditChaveDesbloqueioArquivo: TLabeledEdit;
    Button8: TButton;
    Button10: TButton;
    Button11: TButton;
    Button12: TButton;
    Button13: TButton;
    Button14: TButton;
    procedure Button1Click(Sender: TObject);
    procedure Button2Click(Sender: TObject);
    procedure Button3Click(Sender: TObject);
    procedure Button4Click(Sender: TObject);
    procedure Button5Click(Sender: TObject);
    procedure Button6Click(Sender: TObject);
    procedure Button7Click(Sender: TObject);
    procedure Button10Click(Sender: TObject);
    procedure Button11Click(Sender: TObject);
    procedure Button12Click(Sender: TObject);
    procedure Button13Click(Sender: TObject);
    procedure Button14Click(Sender: TObject);
    procedure Button8Click(Sender: TObject);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  Form1: TForm1;

implementation

{$R *.dfm}

uses Biblioteca;

{$Region 'Controle via Banco'}
procedure TForm1.Button1Click(Sender: TObject);
begin
  ClientDataSet1.Open;
  //
  ClientDataSet1.Append;
  ClientDataSet1.FieldByName('a').AsString := Codifica('C', EditMes.Text);
  ClientDataSet1.FieldByName('b').AsString := Codifica('C', EditAno.Text);
  ClientDataSet1.FieldByName('c').AsString := Codifica('C', EditVencimento.Text);
  ClientDataSet1.FieldByName('d').AsString := Codifica('C', EditBloqueio.Text);
  ClientDataSet1.FieldByName('e').AsString := Codifica('C', EditLiberado.Text);

  ClientDataSet1.FieldByName('f').AsString :=
                                              MD5String(
                                              ClientDataSet1.FieldByName('a').AsString +
                                              ClientDataSet1.FieldByName('b').AsString +
                                              ClientDataSet1.FieldByName('c').AsString +
                                              ClientDataSet1.FieldByName('d').AsString +
                                              ClientDataSet1.FieldByName('e').AsString
                                              );
  ClientDataSet1.Post;
  ClientDataSet1.ApplyUpdates(0);
  //
  SQLTable1.Close;
  SQLTable1.Open;
  ClientDataSet1.Close;
  ClientDataSet1.Open;
end;

procedure TForm1.Button2Click(Sender: TObject);
begin
  SQLTable1.Open;
  //
  ClientDataSet1.Close;
  ClientDataSet1.Open;
end;

procedure TForm1.Button3Click(Sender: TObject);
begin
  Memo1.Clear;
  Memo1.Lines.Add('Mês = ' + Codifica('D', ClientDataSet1.FieldByName('a').AsString));
  Memo1.Lines.Add('Ano = ' + Codifica('D', ClientDataSet1.FieldByName('b').AsString));
  Memo1.Lines.Add('Vencimento = ' + Codifica('D', ClientDataSet1.FieldByName('c').AsString));
  Memo1.Lines.Add('Bloqueio = ' + Codifica('D', ClientDataSet1.FieldByName('d').AsString));
  Memo1.Lines.Add('Liberado = ' + Codifica('D', ClientDataSet1.FieldByName('e').AsString));
end;

procedure TForm1.Button4Click(Sender: TObject);
var
  HashRegistro: String;
begin
  HashRegistro :=
                  MD5String(
                  ClientDataSet1.FieldByName('a').AsString +
                  ClientDataSet1.FieldByName('b').AsString +
                  ClientDataSet1.FieldByName('c').AsString +
                  ClientDataSet1.FieldByName('d').AsString +
                  ClientDataSet1.FieldByName('e').AsString
                  );
  //
  if HashRegistro = ClientDataSet1.FieldByName('f').AsString then
  begin
    Memo1.Lines.Add('===================================');
    Memo1.Lines.Add('Id = ' + ClientDataSet1.FieldByName('id').AsString);
    Memo1.Lines.Add('Hash OK - Registro não violado.');
    Memo1.Lines.Add('===================================');
  end
  else
  begin
    Memo1.Lines.Add('===================================');
    Memo1.Lines.Add('Id = ' + ClientDataSet1.FieldByName('id').AsString);
    Memo1.Lines.Add('Hash Diferente - Registro violado.');
    Memo1.Lines.Add('===================================');
  end;
end;

procedure TForm1.Button5Click(Sender: TObject);
var
  Resultado: boolean;
begin
  // verifica se houve alguma tentativa de burlar os dados no banco de dados
  ClientDataSet1.First;
  while not ClientDataSet1.Eof do
  begin
    Button4.Click;
    ClientDataSet1.Next;
  end;


  // verifica se existe algum registro para o mês e para o ano - caso não exista, bloqueia o sistema.
  // select * from tabela where mes=mes_atual and ano=ano_atual
  ClientDataSet1.First;
  while not ClientDataSet1.Eof do
  begin
    if (YearOf(Date) = StrToInt(Codifica('D', ClientDataSet1.FieldByName('b').AsString))) and (MonthOf(Date) = StrToInt(Codifica('D', ClientDataSet1.FieldByName('a').AsString))) then
    begin
      Resultado := True;
      Break;
    end;
    ClientDataSet1.Next;
  end;
  if not Resultado then
  begin
    ShowMessage('Não encontrou um registro com o mês e o ano atual... Sistema deve ser bloqueado.');
    Exit;
  end;


  // caso tenha encontrado o mes e ano, verifica se a data atual é maior que a data do vencimento
  if DayOf(Date) > StrToInt(Codifica('D', ClientDataSet1.FieldByName('c').AsString)) then
  begin
    ShowMessage('A data atual é maior que a data do vencimento. Será que o cliente já pagou?');

    // verifica se o cliente já realizou o pagamento através do campo Liberado
    if Codifica('D', ClientDataSet1.FieldByName('e').AsString) = 'N' then
    begin
      ShowMessage('O cliente não pagou. Será que o sistema deve ser bloqueado?');

      // caso a data atual seja maior ou igual à data definida para bloqueio, o sistema deve ser bloqueado
      if DayOf(Date) >= StrToInt(Codifica('D', ClientDataSet1.FieldByName('d').AsString)) then
      begin
        ShowMessage('O sistema deve ser bloqueado. Informar o seguinte código para SH: ' + #13#13 + ClientDataSet1.FieldByName('f').AsString);
      end;
    end
    else
      ShowMessage('O cliente já pagou!');
  end;

  // Exercício: e se o cliente deixar de pagar durante dois meses?
end;

procedure TForm1.Button6Click(Sender: TObject);
begin
  Button3.Click;
  //
  Memo1.Lines.Add('=====================================================');
  Memo1.Lines.Add('Hash Registro Bloqueado  = ' + ClientDataSet1.FieldByName('f').AsString);
  Memo1.Lines.Add('Chave Desbloqueio = ' + MD5String(ClientDataSet1.FieldByName('f').AsString + 'uma_chave'));
  Memo1.Lines.Add('=====================================================');
end;

procedure TForm1.Button7Click(Sender: TObject);
begin
  if EditChaveDesbloqueio.Text = MD5String(ClientDataSet1.FieldByName('f').AsString + 'uma_chave') then
  begin
    ClientDataSet1.Edit;

    ClientDataSet1.FieldByName('e').AsString := Codifica('C', 'S');
    ClientDataSet1.FieldByName('f').AsString :=
                                                MD5String(
                                                ClientDataSet1.FieldByName('a').AsString +
                                                ClientDataSet1.FieldByName('b').AsString +
                                                ClientDataSet1.FieldByName('c').AsString +
                                                ClientDataSet1.FieldByName('d').AsString +
                                                ClientDataSet1.FieldByName('e').AsString
                                                );
    ClientDataSet1.Post;
    ClientDataSet1.ApplyUpdates(0);
    //
    SQLTable1.Close;
    SQLTable1.Open;
    ClientDataSet1.Close;
    ClientDataSet1.Open;
    //
    ShowMessage('Dados Atualizados. Sistema Liberado!');
  end
  else
    ShowMessage('Chave não confere - entre em contato com a SH.');
end;
{$EndRegion 'Controle via Banco'}

{$Region 'Controle via Arquivo'}
procedure TForm1.Button10Click(Sender: TObject);
begin
  ClientDataSet2.Append;
  ClientDataSet2.FieldByName('id').AsInteger := ClientDataSet2.RecordCount + 1;
  ClientDataSet2.FieldByName('a').AsString := Codifica('C', EditMes.Text);
  ClientDataSet2.FieldByName('b').AsString := Codifica('C', EditAno.Text);
  ClientDataSet2.FieldByName('c').AsString := Codifica('C', EditVencimento.Text);
  ClientDataSet2.FieldByName('d').AsString := Codifica('C', EditBloqueio.Text);
  ClientDataSet2.FieldByName('e').AsString := Codifica('C', EditLiberado.Text);

  ClientDataSet2.FieldByName('f').AsString :=
                                              MD5String(
                                              ClientDataSet2.FieldByName('a').AsString +
                                              ClientDataSet2.FieldByName('b').AsString +
                                              ClientDataSet2.FieldByName('c').AsString +
                                              ClientDataSet2.FieldByName('d').AsString +
                                              ClientDataSet2.FieldByName('e').AsString
                                              );
  ClientDataSet2.Post;
end;

procedure TForm1.Button11Click(Sender: TObject);
begin
  Memo2.Clear;
  Memo2.Lines.Add('Mês = ' + Codifica('D', ClientDataSet2.FieldByName('a').AsString));
  Memo2.Lines.Add('Ano = ' + Codifica('D', ClientDataSet2.FieldByName('b').AsString));
  Memo2.Lines.Add('Vencimento = ' + Codifica('D', ClientDataSet2.FieldByName('c').AsString));
  Memo2.Lines.Add('Bloqueio = ' + Codifica('D', ClientDataSet2.FieldByName('d').AsString));
  Memo2.Lines.Add('Liberado = ' + Codifica('D', ClientDataSet2.FieldByName('e').AsString));
end;

procedure TForm1.Button12Click(Sender: TObject);
var
  HashRegistro: String;
begin
  HashRegistro :=
                  MD5String(
                  ClientDataSet2.FieldByName('a').AsString +
                  ClientDataSet2.FieldByName('b').AsString +
                  ClientDataSet2.FieldByName('c').AsString +
                  ClientDataSet2.FieldByName('d').AsString +
                  ClientDataSet2.FieldByName('e').AsString
                  );
  //
  if HashRegistro = ClientDataSet2.FieldByName('f').AsString then
  begin
    Memo2.Lines.Add('===================================');
    Memo2.Lines.Add('Id = ' + ClientDataSet2.FieldByName('id').AsString);
    Memo2.Lines.Add('Hash OK - Registro não violado.');
    Memo2.Lines.Add('===================================');
  end
  else
  begin
    Memo2.Lines.Add('===================================');
    Memo2.Lines.Add('Id = ' + ClientDataSet2.FieldByName('id').AsString);
    Memo2.Lines.Add('Hash Diferente - Registro violado.');
    Memo2.Lines.Add('===================================');
  end;
end;

procedure TForm1.Button13Click(Sender: TObject);
var
  Resultado: boolean;
begin
  // verifica se houve alguma tentativa de burlar os dados no banco de dados
  ClientDataSet2.First;
  while not ClientDataSet2.Eof do
  begin
    Button12.Click;
    ClientDataSet2.Next;
  end;


  // verifica se existe algum registro para o mês e para o ano - caso não exista, bloqueia o sistema.
  // select * from tabela where mes=mes_atual and ano=ano_atual
  ClientDataSet2.First;
  while not ClientDataSet2.Eof do
  begin
    if (YearOf(Date) = StrToInt(Codifica('D', ClientDataSet2.FieldByName('b').AsString))) and (MonthOf(Date) = StrToInt(Codifica('D', ClientDataSet2.FieldByName('a').AsString))) then
    begin
      Resultado := True;
      Break;
    end;
    ClientDataSet2.Next;
  end;
  if not Resultado then
  begin
    ShowMessage('Não encontrou um registro com o mês e o ano atual... Sistema deve ser bloqueado.');
    Exit;
  end;


  // caso tenha encontrado o mes e ano, verifica se a data atual é maior que a data do vencimento
  if DayOf(Date) > StrToInt(Codifica('D', ClientDataSet2.FieldByName('c').AsString)) then
  begin
    ShowMessage('A data atual é maior que a data do vencimento. Será que o cliente já pagou?');

    // verifica se o cliente já realizou o pagamento através do campo Liberado
    if Codifica('D', ClientDataSet2.FieldByName('e').AsString) = 'N' then
    begin
      ShowMessage('O cliente não pagou. Será que o sistema deve ser bloqueado?');

      // caso a data atual seja maior ou igual à data definida para bloqueio, o sistema deve ser bloqueado
      if DayOf(Date) >= StrToInt(Codifica('D', ClientDataSet2.FieldByName('d').AsString)) then
      begin
        ShowMessage('O sistema deve ser bloqueado. Informar o seguinte código para SH: ' + #13#13 + ClientDataSet2.FieldByName('f').AsString);
      end;
    end
    else
      ShowMessage('O cliente já pagou!');
  end;

  // Exercício: e se o cliente deixar de pagar durante dois meses?
end;

procedure TForm1.Button14Click(Sender: TObject);
begin
  Button11.Click;
  //
  Memo2.Lines.Add('=====================================================');
  Memo2.Lines.Add('Hash Registro Bloqueado  = ' + ClientDataSet2.FieldByName('f').AsString);
  Memo2.Lines.Add('Chave Desbloqueio = ' + MD5String(ClientDataSet2.FieldByName('f').AsString + 'uma_chave'));
  Memo2.Lines.Add('=====================================================');
end;

procedure TForm1.Button8Click(Sender: TObject);
begin
  if EditChaveDesbloqueioArquivo.Text = MD5String(ClientDataSet2.FieldByName('f').AsString + 'uma_chave') then
  begin
    ClientDataSet2.Edit;

    ClientDataSet2.FieldByName('e').AsString := Codifica('C', 'S');
    ClientDataSet2.FieldByName('f').AsString :=
                                                MD5String(
                                                ClientDataSet2.FieldByName('a').AsString +
                                                ClientDataSet2.FieldByName('b').AsString +
                                                ClientDataSet2.FieldByName('c').AsString +
                                                ClientDataSet2.FieldByName('d').AsString +
                                                ClientDataSet2.FieldByName('e').AsString
                                                );
    ClientDataSet2.Post;
    //
    ShowMessage('Dados Atualizados. Sistema Liberado!');
  end
  else
    ShowMessage('Chave não confere - entre em contato com a SH.');
end;
{$EndRegion 'Controle via Arquivo'}

end.
